### DISCUSSION TITLE
Release workflow feedback / Looking for best practices setting up workspaces

### DISCUSSION THREAD

**Discussion by AlCalzone** [yarnpkg/berry#3067] [Q&A]
I'm currently migrating a monorepo from `yarn` classic to `v3` and trying to drop `lerna` along the way.

The latter is a bit awkward, because there seems to be no easy way to bump changed workspaces and their dependencies to a common version (like `lerna` does in the fixed versioning mode). I've worked around that with a script that...
1. uses `semver` to bump the version defined in my root `package.json`
2. parses the output of `yarn version check` to find changed packages,
3. bumps those workspaces with a call to `yarn workspace <ws> version <new-version> --deferred`
4. repeats 2./3. until there are no more changes
5. finalize the change with `yarn version apply --all` (for some reason `--immediate` "forgets" what I selected for the other packages)
6. creates a git tag and pushes to github, so CI can pick up and do the release

I guess the script could be simplified by making it a yarn plugin and using the API, but that works for now. Maybe a simple command to bump all changed workspaces and their dependencies to a shared version, e.g. `yarn version <strategy> --changed --shared` or something along these lines.

---

However, I'm not sure what's the recommended way to set up the workspace versions and their inter-dependencies. I'm currently using `workspace:` links, but that means the packages cannot be linked into projects that **don't use** `yarn v2`. I should probably be using the actual versions, but then I'm not sure if I need to define the local workspace links using `resolutions` in the root package.json and what happens when those are linked (not necessarily with yarn 2) into an external project.

I think the workspaces docs could use some best practices/examples for this.

**Comment by merceyz** [comment]
For running something on changed workspaces only there is a feature request for that https://github.com/yarnpkg/berry/issues/2374, if that were to be implemented you could do this with
```sh
yarn workspaces foreach --since="<deployed branch>" version <new version> --deferred
yarn version apply --all
```

Until that is implemented https://github.com/yarnpkg/berry/issues/2374#issuecomment-802992606 links to a third party plugin https://github.com/Dcard/yarn-plugins/tree/31803e17fe1c128d86928961bb3007dc2d25a859/packages/changed that you can try using

> I'm currently using workspace: links, but that means the packages cannot be linked into projects that don't use yarn v2

The published package will use the version instead of `workspace:`, but for local links then yeah, that might be problematic.
For projects that do use `yarn@>=2` the [`yarn link`](https://yarnpkg.com/cli/link) command can deal with this for you

**Reply by AlCalzone** [reply]
> links to a third party plugin Dcard/yarn-plugins@31803e1/packages/changed that you can try using

That looked good, but unfortunately that plugin doesn't work in yarn 3 :(
